resolver/upstream_resolver.go

```diff
@@ -169,35 +162,37 @@ func (r *dnsUpstreamClient) fmtURL(ip net.IP, port uint16, _ string) string {
 	return net.JoinHostPort(ip.String(), strconv.Itoa(int(port)))
 }
 
-func (r *dnsUpstreamClient) callExternal(msg *dns.Msg,
-	upstreamURL string, protocol model.RequestProtocol,
+func (r *dnsUpstreamClient) callExternal(
+	ctx context.Context, msg *dns.Msg, upstreamURL string, protocol model.RequestProtocol,
 ) (response *dns.Msg, rtt time.Duration, err error) {
 	if protocol == model.RequestProtocolTCP {
-		response, rtt, err = r.tcpClient.Exchange(msg, upstreamURL)
-		if err != nil {
+		response, rtt, err = r.tcpClient.ExchangeContext(ctx, msg, upstreamURL)
+		if err != nil && r.udpClient != nil {
 			// try UDP as fallback
 			var opErr *net.OpError
-			if errors.As(err, &opErr) && opErr.Op == "dial" && r.udpClient != nil {
-				return r.udpClient.Exchange(msg, upstreamURL)
+			if errors.As(err, &opErr) && opErr.Op == "dial" {
+				return r.udpClient.ExchangeContext(ctx, msg, upstreamURL)
```

Comment:
This changes the behavior: we know have a single timeout for both calls. Meaning if the first one fails because of a timeout, then the second one won't even really start because the context's deadline has been reached.  
I think the best solution would be to start both queries at the same time (or remember which protocol worked last time and use something like weighted random -- but that's more complex and more state).  
That's more work than I want to do in this PR though, so I'll make an issue and get back to it.

In the mean I'll write a test to trigger that condition and do a very ugly thing: ignore the passed-in context for the second query.

Reply:
How about doubling the context timout and splitting it to different contexts for each request? 🤔

Reply:
> How about doubling the context timout and splitting it to different contexts for each request? 🤔

I'd rather not mess with the timeout value the user provides if we can.

When writing the test, I realized this isn't an issue: the only case we retry is in case of an error during "dial", aka connection refused which shouldn't take long so we'll be fine.  
I still think racing TCP and UDP could be worth it and will create an issue.